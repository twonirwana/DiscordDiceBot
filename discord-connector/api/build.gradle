plugins {
    id 'java'
}

group 'de.janno'
version '1.0-SNAPSHOT'

repositories {
    mavenCentral()
}

dependencies {
    implementation "io.projectreactor:reactor-core:3.4.17"
    implementation 'com.google.guava:guava:31.1-jre'

    compileOnly 'org.projectlombok:lombok:1.18.24'
    annotationProcessor 'org.projectlombok:lombok:1.18.22'

    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.2'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.2'
}

java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(14))
    }
}

test {
    useJUnitPlatform()
}