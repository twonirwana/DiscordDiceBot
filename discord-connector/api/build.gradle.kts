plugins {
    id("java")
}

repositories {
    mavenCentral()
}

dependencies {
    implementation(libs.reactor.core)
    implementation(libs.guava)
    implementation(libs.commons.lang3)

    compileOnly(libs.lombok)
    annotationProcessor(libs.lombok)

    testImplementation(libs.junit.jupiter.api)
    testRuntimeOnly(libs.junit.jupiter.engine)
    testImplementation(libs.assertj.core)
}

java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(21))
    }
}

tasks.getByName<Test>("test") {
    useJUnitPlatform()
}