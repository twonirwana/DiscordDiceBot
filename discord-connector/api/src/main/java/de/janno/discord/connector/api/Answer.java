package de.janno.discord.connector.api;

import com.google.common.base.Strings;
import lombok.NonNull;
import lombok.Value;
import reactor.util.annotation.Nullable;

import java.util.List;
import java.util.stream.Collectors;

@Value
public class Answer {
    private static final String MINUS = "\u2212";
    @NonNull
    String title;
    @Nullable
    String content;
    @NonNull
    List<Field> fields;

    public String toShortString() {
        List<String> fieldStringList = fields.stream()
                .map(Field::toShortString)
                .collect(Collectors.toList());

        return String.format("%s %s %s", title,
                        Strings.isNullOrEmpty(content) ? "" : content,
                        fieldStringList.isEmpty() ? "" : fieldStringList.toString())
                .replace("▢", "0")
                .replace("＋", "+")
                .replace(MINUS, "-")
                .replace("*", "");
    }

    @Value
    public static class Field {
        @NonNull
        String name;
        @NonNull
        String value;
        boolean inline;

        public String toShortString() {
            return String.format("%s %s", name, value);
        }
    }
}
