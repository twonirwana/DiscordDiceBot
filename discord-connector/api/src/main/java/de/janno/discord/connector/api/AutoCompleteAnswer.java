package de.janno.discord.connector.api;

import lombok.NonNull;
import lombok.Value;
import org.apache.commons.lang3.StringUtils;

@Value
public class AutoCompleteAnswer {
    @NonNull
    String name;
    @NonNull
    String value;

    public AutoCompleteAnswer(String name, String value) {
        this.name = StringUtils.abbreviate(name, 100);
        this.value = StringUtils.abbreviate(value, 100);
    }
}
