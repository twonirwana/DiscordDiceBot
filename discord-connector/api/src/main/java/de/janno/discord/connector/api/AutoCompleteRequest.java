package de.janno.discord.connector.api;

import lombok.Value;

import java.util.List;

@Value
public class AutoCompleteRequest {
    String focusedOptionName;
    String focusedOptionValue;
    List<OptionValue> optionValues;

}
