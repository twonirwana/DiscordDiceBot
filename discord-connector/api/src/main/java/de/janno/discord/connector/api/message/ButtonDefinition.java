package de.janno.discord.connector.api.message;

import com.google.common.base.Preconditions;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

import java.util.Objects;

@Value
@Builder
public class ButtonDefinition {
    @NonNull
    String label;
    @NonNull
    String id;

    ButtonDefinition.Style style;

    boolean disabled;

    ButtonDefinition(@NonNull String label, @NonNull String id, Style style, boolean disabled) {
        Preconditions.checkArgument(label.length() <= 80, String.format("Label '%s' is to long", label));
        Preconditions.checkArgument(id.length() <= 100, String.format("ID '%s' is to long", id));
        this.label = label;
        this.id = id;
        this.style = Objects.requireNonNullElse(style, Style.PRIMARY);
        this.disabled = disabled;
    }

    public enum Style {
        PRIMARY(1),
        SECONDARY(2),
        SUCCESS(3),
        DANGER(4),
        LINK(5);

        private final int data;

        Style(int i) {
            this.data = i;
        }

        public int getValue() {
            return this.data;
        }
    }
}
