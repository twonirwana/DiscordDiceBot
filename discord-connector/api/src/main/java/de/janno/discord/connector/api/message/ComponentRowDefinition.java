package de.janno.discord.connector.api.message;

import lombok.Builder;
import lombok.Singular;
import lombok.Value;

import java.util.List;

@Value
@Builder
public class ComponentRowDefinition {

    @Singular
    List<ButtonDefinition> buttonDefinitions;
}
