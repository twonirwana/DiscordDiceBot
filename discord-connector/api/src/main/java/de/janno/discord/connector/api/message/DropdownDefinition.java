package de.janno.discord.connector.api.message;

import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import lombok.Builder;
import lombok.Data;
import lombok.NonNull;
import lombok.Value;
import org.apache.commons.lang3.StringUtils;

import java.util.List;

@Value
@Builder
public class DropdownDefinition implements ComponentDefinition {
    String id;
    String placeholder;
    int minValues;
    int maxValues;
    boolean disabled;
    @NonNull
    List<DropdownOption> options;

    DropdownDefinition(String id, String placeholder, int minValues, int maxValues, boolean disabled, @NonNull List<DropdownOption> options) {
        //https://discord.com/developers/docs/interactions/message-components#select-menu-object
        Preconditions.checkArgument(id.length() <= 100, String.format("ID '%s' is to long", id));
        Preconditions.checkArgument(options.size() <= 25, "To many options, max are 25 options");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(id), "id is empty");
        Preconditions.checkArgument(placeholder.length() <= 150, String.format("Placeholder '%s' is to long", placeholder));
        Preconditions.checkArgument(maxValues >= 0 && maxValues <= 25, String.format("maxValue must be between 0 and 25 but was: '%s'", maxValues));
        Preconditions.checkArgument(minValues >= 0 && minValues <= 25, String.format("minValues must be between 0 and 25 but was: '%s'", minValues));
        this.id = id;
        this.placeholder = placeholder;
        this.minValues = minValues;
        this.maxValues = maxValues;
        this.disabled = disabled;
        this.options = options;
    }

    @Override
    public String getLabelOrPlaceholder() {
        return placeholder;
    }

    @Data
    @Builder
    public static class DropdownOption {

        String label;
        @NonNull
        String value;
        String description;
        boolean isDefault;
        String emoji;

        DropdownOption(String label, @NonNull String value, String description, boolean isDefault, String emoji) {
            Preconditions.checkArgument(StringUtils.isNoneBlank(label) || emoji != null, "label and emoji are empty");
            Preconditions.checkArgument(emoji == null || EmojiHelper.isEmoji(emoji), "invalid emoji: " + emoji);
            Preconditions.checkArgument(value.length() <= 100, String.format("value '%s' is to long", value));
            Preconditions.checkArgument(!Strings.isNullOrEmpty(value), "value is empty");
            Preconditions.checkArgument(label == null ||label.length() <= 100, String.format("label '%s' is to long", label));
            Preconditions.checkArgument(description == null || description.length() <= 100, String.format("description '%s' is to long", description));

            this.label = label;
            this.value = value;
            this.description = description;
            this.isDefault = isDefault;
            this.emoji = emoji;
        }
    }
}
