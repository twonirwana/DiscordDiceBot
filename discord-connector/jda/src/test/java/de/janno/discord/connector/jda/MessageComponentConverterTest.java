package de.janno.discord.connector.jda;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.connector.api.message.ButtonDefinition;
import de.janno.discord.connector.api.message.ComponentRowDefinition;
import de.janno.discord.connector.api.message.DropdownDefinition;
import net.dv8tion.jda.api.utils.messages.MessageCreateData;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.List;

@ExtendWith(SnapshotExtension.class)
class MessageComponentConverterTest {

    Expect expect;

    @Test
    void messageComponent2MessageLayout() {
        MessageCreateData messageCreateData = MessageComponentConverter.messageComponent2MessageLayout("welcome", List.of(
                ComponentRowDefinition.builder().componentDefinition(ButtonDefinition.builder()
                        .label("label")
                        .id("id")
                        .emoji("\uD83D\uDE01")
                        .disabled(true)
                        .style(ButtonDefinition.Style.SECONDARY)
                        .build()).build(),
                ComponentRowDefinition.builder().componentDefinition(DropdownDefinition.builder()
                        .id("id")
                        .disabled(true)
                        .maxValues(1)
                        .minValues(1)
                        .placeholder("placeholder")
                        .options(List.of(DropdownDefinition.DropdownOption.builder()
                                .description("description")
                                .emoji("\uD83D\uDE01")
                                .isDefault(true)
                                .label("label")
                                .value("value")
                                .build()))
                        .build()).build()
        ));
        expect.toMatchSnapshot(messageCreateData);
    }
}