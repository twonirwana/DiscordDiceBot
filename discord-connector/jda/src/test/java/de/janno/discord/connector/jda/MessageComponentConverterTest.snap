de.janno.discord.connector.jda.MessageComponentConverterTest.messageComponent2MessageLayout=[
  {
    "content": "welcome",
    "embeds": [ ],
    "files": [ ],
    "components": [
      {
        "components": [
          {
            "id": "id",
            "label": "label",
            "style": "SECONDARY",
            "disabled": true,
            "emoji": {
              "name": "üòÅ"
            },
            "messageCompatible": true,
            "modalCompatible": false
          }
        ],
        "empty": false,
        "enabled": false,
        "valid": true,
        "disabled": true,
        "messageCompatible": true,
        "modalCompatible": false
      },
      {
        "components": [
          {
            "id": "id",
            "placeholder": "placeholder",
            "minValues": 1,
            "maxValues": 1,
            "disabled": true,
            "options": [
              {
                "label": "label",
                "value": "value",
                "description": "description",
                "isDefault": true,
                "emoji": {
                  "name": "üòÅ"
                },
                "default": true
              }
            ],
            "messageCompatible": true,
            "modalCompatible": false
          }
        ],
        "empty": false,
        "enabled": false,
        "valid": true,
        "disabled": true,
        "messageCompatible": true,
        "modalCompatible": false
      }
    ],
    "mentions": {
      "mentionParse": [
        "USER",
        "ROLE",
        "CHANNEL",
        "EMOJI",
        "HERE",
        "EVERYONE",
        "SLASH_COMMAND"
      ],
      "mentionUsers": [ ],
      "mentionRoles": [ ],
      "mentionRepliedUser": true
    },
    "tts": false,
    "flags": 4096,
    "suppressedNotifications": true,
    "mentionRepliedUser": true,
    "voiceMessage": false,
    "suppressEmbeds": false
  }
]